### PR TITLE
Add verification support on cva6_reorg branch to 

### PR DESCRIPTION
**Pull request by ASintzoff**
- support of rfvi
- support of vcs
- support of vsim
- support of FPGA

Note this PR is strongly linked to the corresponding PR on core-v-verif repository.

Modifications done here are needed to have cva6 verification provided in core-v-verif working after the cva6 reorganization.

### CODE REVIEW

File: core/include/riscv_pkg.sv
Comment (MikeOpenHWGroup) [inline core/include/riscv_pkg.sv:31]: Is there a way we can do this without using compiler macros (`defines)?

Comment (MikeOpenHWGroup) [inline core/include/riscv_pkg.sv:31]: Is there a way we can do this without compiler macros (`defines)?

**Review by MikeOpenHWGroup — COMMENTED** [review]
This looks fine to me @ASintzoff, thanks for this!  I have made a couple of review comments, but these are for discussion, they do not need to gate the merge.   Having said that, I will not approve the merge since this should come from @zarubaf.

File: .gitmodules
Comment (ASintzoff) [reply .gitmodules:45]: I would prefer to not fork if possible.
IMO, it seems better to modify riscv-dbg according to https://github.com/pulp-platform/riscv-dbg/pull/83

File: core/include/riscv_pkg.sv
Comment (ASintzoff) [reply core/include/riscv_pkg.sv:31]: Good point. @JeanRochCoulon will address it. Details will follow.

Comment (JeanRochCoulon) [reply core/include/riscv_pkg.sv:31]: XLEN is defined in a package. As there is no way to override a parameter in a package, we can create different versions of the same package and choose which version to compile. By the way, Flist.cva6 should be duplicated into Flist.cv32a6 and Flist.cv64a6. In that way, all the supported configurations (to be referred to cva6 requirement specification) would correspond to a dedicated FLIST file as FLIST.cva64a6_noFD for instance (naming to be fine-tuned of course).

@zarubaf @MikeOpenHWGroup, I can't wait for having your feedback !?!

File: .gitmodules
Comment (zarubaf) [reply .gitmodules:45]: I think @JeanRochCoulon and I agree that the problem he was seeing because of a bug in the debug unit that should have been fixed by later commits. Can you try updating to the latest `riscv-dbg` module? We need to merge this before we can merge this PR as the public CVA6 can't depend on "private" forks of members.

File: Makefile
Comment (zarubaf) [inline Makefile:39]: Why is this removed, isn't it working correctly?

File: core/ariane.sv
Comment (zarubaf) [inline core/ariane.sv:45]: I would prefer structs on the top ports. Also, I guess we want to make the inclusion of the tracer port optional to remain backward compatibility.

File: core/ariane.sv
Comment (zarubaf) [inline core/ariane.sv:95]: Strongly prefer the struct variant. Why would you unroll?

File: corev_apu/tb/ariane_testharness.sv
Comment (zarubaf) [inline corev_apu/tb/ariane_testharness.sv:823]:
```suggestion
  );
```

File: core/include/riscv_pkg.sv
Comment (zarubaf) [reply core/include/riscv_pkg.sv:31]: Technically you could fork that out into two separate packages `cv64a_config_pkg.sv` and `cv32a_config_pkg.sv` and include the one that you are looking for.

Comment (zarubaf) [reply core/include/riscv_pkg.sv:31]: My feeling is that, at a certain point, we are better off if we generate part of the RTL from templates as we are hitting on the limits of SV.

**Review by zarubaf — CHANGES_REQUESTED** [review]
Generally looks okay to me:

- We need to get the debug thing sorted.
- CI is failing
- I really don't like the unrolled interfaces, what is the rationale?
- Small nits

File: Makefile
Comment (ASintzoff) [reply Makefile:39]: In 32-bit configuration, tests in riscv-tests and riscv-compliance test suites are randomly stucked. Several executions of the same test are stopping at different places. Stopping means no more instruction logged in the trace but `cycles` still incremented.

File: corev_apu/tb/ariane_testharness.sv
Comment (ASintzoff) [reply corev_apu/tb/ariane_testharness.sv:823]: Good point

File: .gitmodules
Comment (ASintzoff) [reply .gitmodules]: With latest version of `riscv-dbg`, the problem disappears. PR has been adapted accordingly.

File: core/ariane.sv
Comment (JeanRochCoulon) [reply core/ariane.sv:45]: Florian,

When structure are used in RTL, dc_shell from Synopsys flattens the structure by generating a logic bit vector. A signal referring to rvfi_inst_t typedef will be mapped by dc_shell on a single logic bit vector of more than 100 bits. While it could remain transparent when the signal is internal, it become a problem when it is in IO.

For instance, to execute a gate simulation, we replace the ariane RTL module by the ariane gate module, but testharness’s RVFI and AXI structures cannot be connected to the ariane’s two logic bit vectors.

I ask for a command line option to the Synopsys support, but without success.

We already get the same problem, a long time ago on LEON Sparc. We had flattened the structures.

My proposal is to expand as minimal the ariane input and output which prevent us from executing gate simulation connected to the testharness, that’s to say RVFI and AXI.

Regards

File: core/include/riscv_pkg.sv
Comment (JeanRochCoulon) [reply core/include/riscv_pkg.sv]: As you will see, a RTL proposal has been pushed.  A new package has been created to host the cva6 configuration. Two Flist.cva6xxx files have been created, one to point on the package file where XLEN=64 and the other to the package file with XLEN=32. According to the variant, the Makefile selects the right Flist.cva6xxx file. Tell me if it works for you! Cheers

File: core/ariane.sv
Comment (zarubaf) [reply core/ariane.sv:45]: I understand your use-case, then let's add a wrapper around the main `ariane.sv` (`ariane_wrapepr.sv`?) that exposes the interfaces as flat ports. In that way, we do not break the interface for existing users.

File: core/ariane.sv
Comment (JeanRochCoulon) [reply core/ariane.sv]: @zarubaf    As discussed, the structures generate issues at dc_shell synthesis time. To solve the problem, a wrapper has been added in core testbench (testbench used for UVM) following the suggestion, testharness is kept unchanged.

File: core/Flist.cv64a6_imacfd_sv39
Comment (zarubaf) [inline core/Flist.cv64a6_imacfd_sv39:1]: Super nitpick: The isa string is `imafdc` not `imacfd`

File: core/ariane.sv
Comment (zarubaf) [reply core/ariane.sv:45]: Adding an interface to the SystemVerilog module breaks its interface so we should ifdef similar to the firesim trace port.

File: core/include/cv32a6_imac_sv0_config_pkg.sv
Comment (zarubaf) [inline core/include/cv32a6_imac_sv0_config_pkg.sv:13]: Nit: According to the style guide localparams are supposed to be CamelCase.
```suggestion
    localparam CVA6ConfigXlen = 32;
```

File: core/include/cv64a6_imacfd_sv39_config_pkg.sv
Comment (zarubaf) [inline core/include/cv64a6_imacfd_sv39_config_pkg.sv:13]: Same nit here

**Review by zarubaf — COMMENTED** [review]
Almost there.

File: core/Flist.cv64a6_imacfd_sv39
Comment (JeanRochCoulon) [reply core/Flist.cv64a6_imacfd_sv39:1]: Yes, super nitpick!!!! :-)     Fixed by commit

File: core/include/cv32a6_imac_sv0_config_pkg.sv
Comment (JeanRochCoulon) [reply core/include/cv32a6_imac_sv0_config_pkg.sv]: Fixed by commit

File: core/include/cv64a6_imacfd_sv39_config_pkg.sv
Comment (JeanRochCoulon) [reply core/include/cv64a6_imacfd_sv39_config_pkg.sv]: Fixed by commit

**Comment by JeanRochCoulon** [comment]
@zarubaf    Concerning the RFVI interface add-on, I am confused: I though @MikeOpenHWGroup would like to limit/remove all the ifdef from CVA6 ?!  Mike, could you confirm it?

File: core/include/riscv_pkg.sv
Comment (zarubaf) [inline core/include/riscv_pkg.sv:18]: Let's move that into the package. It otherwise pollutes the namespace

**Comment by zarubaf** [comment]
> @zarubaf Concerning the RFVI interface add-on, I am confused: I though @MikeOpenHWGroup would like to limit/remove all the ifdef from CVA6 ?! Mike, could you confirm it?

What we can't change is the top-level interface. (well we can but that would mean a major version bump - that is not necessary imho). Other approach would be to promote `ariane.sv` to a wrapper around `ariane_v2.sv` which ties off the ports - ugly. Imho we should go with the ifdef.

File: core/include/riscv_pkg.sv
Comment (JeanRochCoulon) [reply core/include/riscv_pkg.sv]: done

**Comment by JeanRochCoulon** [comment]
Ok, understood, thanks for the clarification, I have added a RVFI_TRACE ifdef in ariane module.

**Comment by JeanRochCoulon** [comment]
@zarubaf  @MikeOpenHWGroup   Very good, we have solved the riscv-dbg issue, managed the different configurations (will be useful), and added RVFI. Splendid.  
Asap the merge is done, we tranfer Thales guys (and who wants) on the new organization.  Cheers

File: corev_apu/fpga/src/ariane_xilinx.sv
Comment (zarubaf) [inline corev_apu/fpga/src/ariane_xilinx.sv:436]: Wait, this should be reverted to the struct-based version.

**Review by zarubaf — APPROVED** [review]
Just the small FPGA fix and then let's merge!

File: corev_apu/fpga/src/ariane_xilinx.sv
Comment (ASintzoff) [reply corev_apu/fpga/src/ariane_xilinx.sv]: Modification removed from commit Support FPGA generation

**Review by zarubaf — APPROVED** [review]
Thanks for all the work. LGTM - merging.

**Pull request merged**
